File: 0322-coin-change/0322-coin-change.cpp
class Solution {
public:
    int arr[10001];
    int fun(vector<int>& coins, int amount){
        if(amount == 0){
            return 0;
        }
        
        
        int i = INT_MAX;
        for(auto x: coins){
            if(amount - x < 0){
                   continue;
            }    
            int k;
            if(arr[amount - x] == -1){    
                    k = fun(coins, amount - x);
            }
            else{
                    k = arr[amount - x];
            }
            if(k != INT_MAX){
                i = min(i, k+1);
            }
        }
        
        arr[amount] = i;
        return i;        
    }
    
    int coinChange(vector<int>& coins, int amount) {
        memset(arr, -1, sizeof(arr));
        int ans = fun(coins, amount);
        if(ans == INT_MAX) return -1;
        return ans; 
    }
};

File: 0322-coin-change/0322-coin-change.java
class Solution {

    private int[][] dp = new int[99999][];

    {
        for(int i = 0; i < 99999; i += 1){
            dp[i] = new int[12];
            for(int j = 0; j < 12; j += 1){
                dp[i][j] = -2;
            }
        }
    }

    public int coinChange(int[] coins, int amount) {
        int n = coins.length;
        return helper(amount, 0, coins, n);
    }

    private int helper(int amount, int idx, int[] coins, int n){
        if(amount < 0){
            return -1;
        }

        if(amount == 0){
            return 0;
        }

        if(idx == n){
            return -1;
        }

        if(dp[amount][idx] != -2){
            return dp[amount][idx];
        }

        // take
        int step1 = helper(amount - coins[idx], idx, coins, n);
        
        //not take
        int step2 = helper(amount, idx + 1, coins, n);

        if (step1 == -1 && step2 == -1) {
            return dp[amount][idx] = -1;
        }
        else if (step1 == -1) {
            // don't count
            return dp[amount][idx] = step2;
        } else if (step2 == -1) {
            return dp[amount][idx] = step1 + 1;
        }
        else {
            if (step1 < step2) {
                return dp[amount][idx] = step1 + 1;
            } else {
                return dp[amount][idx] = step2;
            } 
        }
    }
}